### PR TITLE
RET-2411: ET3 Response Form doc generation

### PR DESCRIPTION
**Pull request by JackReevies**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RET-2411

### Change description ###

Added ET3 Response Form doc generation

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/domain/documents/Et3ResponseData.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+import lombok.Data;
+import lombok.experimental.SuperBuilder;
+
+import java.util.List;
+
+@Data
+@SuperBuilder
+@SuppressWarnings({"PMD.LinguisticNaming", "PMD.TooManyFields"})
+public class Et3ResponseData {
```
Comment (nickwest1) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/domain/documents/Et3ResponseData.java:11]: Although it's a simple POJO, it should still have JavaDoc!
Explain a bit about what it's used for

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/Et3ResponseHelper.java [hunk 2]
```diff
@@ -17,9 +22,14 @@
...
 public class Et3ResponseHelper {
 
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private static final String TEMPLATE_NAME = "ET3_0922 (1).docx";
+    private static final String OUTPUT_NAME = "ET3 Response.pdf";
     private static final String CLAIMANT_NAME_TABLE = "<pre> ET1 claimant name&#09&#09&#09&#09 %s</pre><hr>";
     public static final String START_DATE_MUST_BE_IN_THE_PAST = "Start date must be in the past";
     public static final String END_DATE_MUST_BE_AFTER_THE_START_DATE = "End date must be after the start date";
+    public static final String CHECKED = "■";
+    public static final String UNCHECKED = "□";
```
Comment (nickwest1) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/Et3ResponseHelper.java:32]: Unusual characters here that are worth a comment on the right with the unicode character code for developer clarity

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/Et3ResponseHelper.java [hunk 3]
```diff
@@ -67,4 +77,140 @@ public static List<String> validateEmploymentDates(CaseData caseData) {
...
+            .et3ResponsePayTakehome(caseData.getEt3ResponsePayTakehome())
+            .et3ResponseCorrectNoticeDetails(caseData.getEt3ResponseCorrectNoticeDetails())
+            .et3ResponseContestClaimDetails(caseData.getEt3ResponseContestClaimDetails())
+            .et3ResponseEmployerClaimDetails(caseData.getEt3ResponseEmployerClaimDetails())
+            .et3ResponseRespondentSupportDetails(caseData.getEt3ResponseRespondentSupportDetails())
+            .et3ResponsePensionCorrectDetails(caseData.getEt3ResponsePensionCorrectDetails())
+            .hearingPhone(caseData.getEt3ResponseHearingRespondent().contains("Phone hearings") ? "■" : "□")
+            .hearingVideo(caseData.getEt3ResponseHearingRespondent().contains("Video hearings") ? "■" : "□")
+            .repHearingPhone(caseData.getEt3ResponseHearingRepresentative().contains("Phone hearings") ? "■" : "□")
+            .repHearingVideo(caseData.getEt3ResponseHearingRepresentative().contains("Video hearings") ? "■" : "□")
```
Comment (nickwest1) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/Et3ResponseHelper.java:166]: Get these characters from constants, which you've already created above anyway

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseService.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.ecm.common.exceptions.DocumentManagementException;
+import uk.gov.hmcts.et.common.model.ccd.CaseData;
+import uk.gov.hmcts.et.common.model.ccd.DocumentInfo;
+
+@Slf4j
+@Service("et3ResponseService")
+@RequiredArgsConstructor
+public class Et3ResponseService {
```
Comment (nickwest1) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseService.java:13]: JavaDoc?

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/Et3ResponseHelper.java [hunk 5]
```diff
@@ -17,9 +22,14 @@
     "PMD.ExcessiveMethodLength", "PMD.ClassNamingConventions", "PMD.PrematureDeclaration"})
 public class Et3ResponseHelper {
 
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private static final String TEMPLATE_NAME = "ET3_0922 (1).docx";
```
Comment (hsjhita) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/Et3ResponseHelper.java:26]: Template naming should be in a specific format - we can get into a call and discuss this

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/Et3ResponseHelperTest.java [hunk 6]
```diff
@@ -93,4 +101,54 @@ void givenStartDateInTheFutureAndEndDateBeforeStartDate_shouldReturnErrors() {
...
+        caseData.setEt3ResponseContinuingEmployment("No");
+        caseData.setEt3ResponseIsPensionCorrect("No");
+        caseData.setEt3ResponseContactPreference("Post");
+        caseData.setEt3ResponsePayFrequency("Weekly");
+
+        String expected = new String(Files.readAllBytes(Paths.get(Objects.requireNonNull(getClass().getClassLoader()
+            .getResource("et3ResponseDocument.json")).toURI())));
+
+        String actual = Et3ResponseHelper.getDocumentRequest(caseData, "any");
+        assertThat(actual, is(expected));
```
Comment (nickwest1) [inline src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/Et3ResponseHelperTest.java:152]: Minor:  Better to do it like this from a readability perspective:
```
assertThat(actual)
    .isEqualTo(expected);

```

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        doCallRealMethod().when(documentManagementService).addDocumentToDocumentField(documentInfo);
+    }
+
+    @Test
+    void generateEt3ProcessingDocument() throws IOException {
+        when(tornadoService.generateEventDocument(any(CaseData.class), anyString(),
+            anyString(), anyString())).thenReturn(documentInfo);
+        DocumentInfo documentInfo1 = et3ResponseService.generateEt3ResponseDocument(new CaseData(), "userToken",
+            ENGLANDWALES_CASE_TYPE_ID);
+        assertThat(documentInfo1, is(documentInfo));
```
Comment (nickwest1) [inline src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseServiceTest.java:61]: As above RE assertThat

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        when(tornadoService.generateEventDocument(any(CaseData.class), anyString(),
+            anyString(), anyString())).thenThrow(new InternalException(ERROR_MESSAGE));
+        assertThrows(Exception.class, () -> et3ResponseService.generateEt3ResponseDocument(new CaseData(), "userToken",
+            ENGLANDWALES_CASE_TYPE_ID));
+    }
+
+    @Test
+    void assertThatEt3DocumentIsSaved() {
+        et3ResponseService.saveEt3ResponseDocument(caseData, documentInfo);
+        assertThat(caseData.getEt3ResponseDocument().getDocumentFilename(), is(documentInfo.getDescription()));
```
Comment (nickwest1) [inline src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseServiceTest.java:75]: As above RE assertThat

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/Et3ResponseHelper.java [hunk 9]
```diff
@@ -17,9 +22,14 @@
...
+    private static final String OUTPUT_NAME = "ET3 Response.pdf";
```
Comment (hsjhita) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/Et3ResponseHelper.java:27]: Just wondering - if we had multiple respondents, would we be better off appending the respondents name to the output name? e.g Jack Reeve - ET3 Response?

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/utils/CaseDataBuilder.java [hunk 10]
```diff
@@ -348,7 +348,7 @@ public CaseDataBuilder withRespondentWithAddress(String respondentName, String a
         return this;
     }
 
-    private Address createAddress(String addressLine1, String addressLine2, String addressLine3,
+    public Address createAddress(String addressLine1, String addressLine2, String addressLine3,
```
Comment (nickwest1) [inline src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/utils/CaseDataBuilder.java:351]: JavaDoc?

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/domain/documents/Et3ResponseData.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
+package uk.gov.hmcts.ethos.replacement.docmosis.domain.documents;
+
+import lombok.Data;
+import lombok.experimental.SuperBuilder;
+
+import java.util.List;
+
```
Comment (hsjhita) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/domain/documents/Et3ResponseData.java:7]: Java Doc

File: src/test/resources/et3ResponseDocument.json [hunk 12]
```diff
@@ -0,0 +1 @@
...upportNeeded":null,"et3ResponseRespondentSupportDetails":"support details","titleMr":"■","titleMrs":"□","titleMiss":"□","titleMs":"□","titleOther":"□","titleText":null,"contactPost":"■","contactEmail":"□","hearingVideo":"□","hearingPhone":"■","siteYes":"■","siteNo":"□","acasYes":"□","acasNo":"■","datesYes":"□","datesNo":"■","datesNA":"□","continueYes":"□","continueNo":"■","continueNA":"□","jobYes":"□","jobNo":"■","jobNA":"□","hoursYes":"□","hoursNo":"■","hoursNA":"□","earnYes":"□","earnNo":"■","earnNA":"□","beforeWeekly":"■","beforeMonthly":"□","beforeAnnually":"□","takehomeWeekly":"■","takehomeMonthly":"□","takehomeAnnually":"□","noticeYes":"□","noticeNo":"□","noticeNA":"■","pensionYes":"□","pensionNo":"■","pensionNA":"□","contestYes":"■","contestNo":"□","eccYes":"□","helpYes":"■","helpNo":"□","helpNA":"□","repName":null,"repOrgName":null,"repAddressLine1":null,"repAddressLine2":null,"repTown":null,"repCounty":null,"repPostcode":null,"repDXNumber":null,"repPhoneNumber":null,"repMobileNumber":null,"repContactEmail":"□","repContactPost":"□","repEmailAddress":null,"repHearingVideo":"■","repHearingPhone":"□","et3ResponseRespondentMobile":null,"et3ResponseRespondentEmailAddress":null}}
```
Comment (nickwest1) [inline src/test/resources/et3ResponseDocument.json:1]: For these checked/unchecked characters, better we use a constant.  Only that I can see all kinds of problems coming in later if we don't

**Review by nickwest1 — CHANGES_REQUESTED** [review]
Good PR in general, just a bunch of small bits here

File: src/test/resources/et3ResponseDocument.json (hunk 12 above)
Comment (JackReevies) [reply src/test/resources/et3ResponseDocument.json:1]: Is that do-able in this json file? It's only read in as a string for a test

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/utils/CaseDataBuilder.java (hunk 10 above)
Comment (JackReevies) [reply src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/utils/CaseDataBuilder.java:351]: oops my mistake this should be private still

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/Et3ResponseHelperTest.java (hunk 6 above)
Comment (JackReevies) [reply src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/Et3ResponseHelperTest.java:152]: Where does this `assertThat` come from? This file is currently importing from `org.hamcrest.MatcherAssert.assertThat` and it doesnt recognise an `assertThat` that takes a single argument

**Review by hsjhita — APPROVED** [review]

**Pull request merged**